File: exp-10/q1-bfs/solution.js
const readline = require('readline');

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

let lines = [];
rl.on('line', line => lines.push(line.trim()));
rl.on('close', () => {
    const data = lines.join(' ').split(/\s+/).map(Number);
    let idx = 0;
    const n = data[idx++], m = data[idx++];
    const adj = Array.from({ length: n + 1 }, () => []);
    for (let i = 0; i < m; i++) {
        const u = data[idx++], v = data[idx++];
        adj[u].push(v);
        adj[v].push(u);
    }
    const s = data[idx++];

    for (let i = 1; i <= n; i++) adj[i].sort((a, b) => a - b);

    const visited = new Array(n + 1).fill(false);
    const queue = [s];
    visited[s] = true;
    const order = [];
    while (queue.length) {
        const u = queue.shift();
        order.push(u);
        for (const v of adj[u]) {
            if (!visited[v]) {
                visited[v] = true;
                queue.push(v);
            }
        }
    }
    console.log('BFS Traversal: ' + order.join(' '));
});

File: exp-10/q1-bfs/solution.py
from collections import deque

n, m = map(int, input().split())
adj = [[] for _ in range(n + 1)]
for _ in range(m):
    u, v = map(int, input().split())
    adj[u].append(v)
    adj[v].append(u)
s = int(input())

for i in range(1, n + 1):
    adj[i].sort()

visited = [False] * (n + 1)
q = deque([s])
visited[s] = True

order = []
while q:
    u = q.popleft()
    order.append(u)
    for v in adj[u]:
        if not visited[v]:
            visited[v] = True
            q.append(v)

print("BFS Traversal:", " ".join(map(str, order)))